### DISCUSSION TITLE
I want to open the air app directly in the web url.

### DISCUSSION THREAD

**Discussion by a9848840** [airsdk/Adobe-Runtime-Support#1735] [Q&A] [answered]
I want to open the air app directly in the web url.
Do I have to add ane?

The ane below looks packaged
So adding ane can start the air app directly from the web url?
https://www.myflashlabs.com/open-adobe-air-app-browser-pass-parameters/

This is the setting of AndroidManifest.xml in the android calling app
https://stackoverflow.com/questions/2958701/launch-custom-android-application-from-android-browser/2958870#2958870

I tried adding in application.xml

<activity android:hardwareAccelerated="false">
<intent-filter>
<action android:name="android.intent.action.MAIN" />
<category android:name="android.intent.category.LAUNCHER" />
</intent-filter>
<intent-filter>
<action android:name="android.intent.action.VIEW" />
<category android:name="android.intent.category.BROWSABLE" />
<category android:name="android.intent.category.DEFAULT" />
<data android:scheme="air.myapp" />
</intent-filter>
</activity>
![image](https://user-images.githubusercontent.com/39514485/156490505-167078cd-9476-4edb-88bb-46edb85ff950.png)

"air.myapp" is my app PackageName,
Web Url  air.myapp://https://www.*******.com.tw/
But still can't open the app

Is ane necessary to implement this functionality?

**Comment by marchbold** [comment, marked as answer]
You need to customise the main activity in your application.  **You do not need an ANE for this on Android**, just add the following `activity` to your application descriptor in your manifest additions and replace `yoururlscheme` appropriately:

```xml
<android>
	<manifestAdditions><![CDATA[
		<manifest android:installLocation="auto" >
			<application>


				<activity>
					<intent-filter>
						<action android:name="android.intent.action.MAIN"/>
						<category android:name="android.intent.category.LAUNCHER"/>
					</intent-filter>
					<intent-filter>
						<action android:name="android.intent.action.VIEW"/>
						<category android:name="android.intent.category.DEFAULT"/>
						<category android:name="android.intent.category.BROWSABLE"/>
						<data android:scheme="yoururlscheme"/>
					</intent-filter>
				</activity>


			</application>
		</manifest>
	]]></manifestAdditions>
</android>
```

Note the placement of the activity inside the application tag in your manifest additions. Also I don't believe url schemes can have dots in them?

**Comment by a9848840** [comment]
@marchbold 

May I ask you, how to set the IOS XML?
Or just adjust the WEB URL?
to open the air ios app using the web url


thanks

**Reply by marchbold** [reply]
Sure, the iOS one is as follows:

```xml
<iPhone>
	<InfoAdditions><![CDATA[

		<key>CFBundleURLTypes</key>
		<array>
			<dict>
				<key>CFBundleURLSchemes</key>
				<array>
					<string>yoururlscheme</string>
				</array>
			</dict>
		</array>

	]]></InfoAdditions>
</iPhone>
```

Note: `CFBundleURLTypes` is an array so if you have multiple values you need to replicate the `dict` node in the array. 

It's worth noting that on iOS Apple recommends ensuring you validate all parameters etc as this has been used as an attack vector with iOS though likely this applies to any platform. Apple recommend usage of universal links now.

https://developer.apple.com/documentation/xcode/defining-a-custom-url-scheme-for-your-app

**Reply by a9848840** [reply]
Thank you very much for your assistance
android has been tested successfully
Next continue ios

**Reply by a9848840** [reply]
@marchbold 

I have doubts in adding ios app xml



> 				<key>CFBundleURLTypes</key>
> 					<array>
> 						<dict>
> 							<key>CFBundleTypeRole</key>
> 								<string>Editor</string>
> 							<key>CFBundleURLSchemes</key>
> 								<array>
> 									<string>line3rdp.myapp</string>
> 									<string>fbmyfb</string>
> 								</array>
> 						</dict>
> 					</array>


I try to
`<string>fbmyfb</string>`
Add the following line
`<string>myapp</string>`

> 								<array>
> 									<string>line3rdp.myapp</string>
> 									<string>fbmyfb</string>
> 									<string>myapp</string>
> 								</array>

i found myapp cant’t = BundleID

thank you!!!

**Reply by marchbold** [reply]
Yes, you can't use the bundle id. 

Probably want something like this:

```xml
<iPhone>
	<InfoAdditions><![CDATA[

		<key>CFBundleURLTypes</key>
		<array>
			<dict>
				<key>CFBundleTypeRole</key>
				<string>Editor</string>
				<key>CFBundleURLSchemes</key>
				<array>
					<string>line3rdp.myapp</string>
					<string>fbmyfb</string>
				</array>
			</dict>
			<dict>
				<key>CFBundleURLSchemes</key>
				<array>
					<string>yoururlscheme</string>
				</array>
			</dict>
		</array>

	]]></InfoAdditions>
</iPhone>
```

**Reply by marchbold** [reply]
Also note:

> A custom scheme must begin with an alphabetic character (A–Z, a–z) and may be followed by any number of alphanumeric characters, +, -, or . It also cannot be any of the following: "javascript", "vbscript", "data", "blob", "http", "https", "mailto", "livescript", "facetime", "facetime-audio".

**Reply by a9848840** [reply]
Thanks for the reminder, and the details

Set the scheme in ios xml
`<string>air.myBunndleID</string>`

On the web side set to

`<a href="air.myBunndleID://https://www.******.com.tw/">OpenYourApp ****** IOS 1.7</a></br>`


safari can't open the page because the url is invalid
Is this the case on the web side?

thank you ><

**Reply by marchbold** [reply]
That url doesn't look valid, you should url encode the parameters and do something like: 

```
air.myBundleID://open?url=https%3A%2F%2Fwww.%2A%2A%2A%2A%2A%2A.com.tw%2F
```

**Reply by a9848840** [reply]
Thank you for your help
ios test succeeded
^^

**Reply by a9848840** [reply]
I want to ask you again
I need to open the APP through the link and execute a specific function. Where can I carry the parameters? Does the APP need a place to receive the parameters? To execute the subsequent functions

thanks

**Reply by marchbold** [reply]
You should be able to receive these through the invoke event
